docs/rfc-1-staking-contract.adoc

```diff
@@ -0,0 +1,381 @@
+
+== Specification
+
+=== Functionality
+
+==== Delegating a stake
+
+Tokens are delegated by the owner. During the delegation, the owner needs to
+appoint an operator, beneficiary, and authorizer. The owner may decide to
+delegate just a portion of their tokens. The owner may delegate multiple times
+to different operators. For simplicity, the operator address can not be reused
+between delegations even if the delegation has been recovered. The only
+exception is when the delegation has been canceled. To protect against griefing
+by frontrunning stake delegation transactions and reserving operator addresses,
+staking contract should require a governable minimum stake amount for each
+delegation.
+
+Staking contract binds the owner with the operator, beneficiary, and authorizer
+for the given stake delegation and records the stake delegation amount.
+Delegation is set to active after 2 hours, once the cancellation window passes.
+
+Existing KEEP and NU stake owners also need to perform a delegation so that T
+staking contract can record their operator, beneficiary, and authorizer. Since
+these values are read-only, they should be copied from KEEP staking contract to
+T staking contract and NU stakers should set them during the delegation. For
+existing stakes, the entire stake amount is registered in T staking contract.
+
+There are 2 hours to cancel a stake without suffering an undelegation period.
+This functionality is needed e.g. in case of a typo in the operator address.
+Undelegation can not be canceled in the same transaction in which it was
+created. Once the delegation has been canceled, the operator address can be used
+again.
+
+==== Authorizing an application
+
+Before the operator is eligible to participate in the given application, the
+authorizer appointed during the stake delegation needs to review the application
+and approve it to use the stake. From the moment of approval, the application
+is authorized to slash or seize the stake.
+
+The authorizer decides up to which amount the given application is eligible to
+slash. This allows managing the risk profile of the staker individually for each
+application. The authorized stake amount can only be increased.
+
+Authorization can not be canceled.
+
+The application being authorized is notified via a callback every single time
+new authorization happens and the application may decide to revert the
+transaction if individual requirements of the application are not met. Such
+requirements could be, for example, not enough stake authorized for the
+particular application or not supported legacy stake (see Limitations section).
+
+All applications first need to be authorized in a registry of T applications
+controlled by the governance.
+
+==== Stake top-up
+
+Top-up increases the amount of tokens locked in a delegation. This increases the
+probability of being chosen for work in the network but is only effective for
+future checks of the stake amount.
```

Comment:
Looks like `This increases the probability of being chosen for work in the network but is only effective for
future checks of the stake amount.` conflicts with `Stake top-up does not automatically increase authorization levels for applications.`

Reply:
Do you mean that the top-up does not automatically increase this possibility given that the higher amount first needs to be authorized? If so, agreed. And I'll clarify it.